### DISCUSSION TITLE
Reflection

### DISCUSSION THREAD

**Discussion by web-fu** [azjezz/psl#402] [Ideas]
Lately I'm working on a mapping library and I would like to add annotation support.
I used PHP Native Reflection, but it is very inconsistent.
Many functions return `false` instead of `null`, there is no way to get annotation directly, fetching types from properties, parameters and return is confusing...
I saw that some libraries (PHPStan, for example) use their own Reflection implementation, so my question is: why there is not a dedicated library? 
I'm not sure the scope of PSL project is to include reflection, but I think a standalone standardized tool would be very usefull.
What is your opinion?

**Comment by veewee** [comment]
Hello,

Have you looked into https://github.com/Roave/BetterReflection ?

I think it solves most of the issues you mentioned.
Currently, I don't have any strong opinions on having a reflection layer in psl.

**Reply by web-fu** [reply]
Yes, I love it.
But quoting the library:

> Better Reflection is NOT suited to runtime usage, since performance is much worse than PHP built-in reflection. If you do not want to do anything that native PHP reflection can't do, then just use native PHP reflection! The "Better" in Better Reflection refers to feature, not speed!

Maybe it's possible to have a coherent library more runtime friendly
Just having annotation support would be great

**Comment by azjezz** [comment]
I like the idea of having a reflection API in PSL, that is better typed, it seems all good in theory, but implementing something that would be truly type safe is not really possible ( or at least not easy ).

I'll be open to suggestions of how such an API would look like, then we can discuss it further.

**Reply by web-fu** [reply]
I'm not sure what could be the effort of making the Reflection API type safe.
Yesterday I tried encapsulating a couple of classes:
https://github.com/web-fu/reflection
The first obstacle in PHPStan is in the constructor definition:
>  ------ ----------------------------------------------------------------------------------------------------------------
>  Line   src/Reflection/ReflectionClass.php
> ------ ----------------------------------------------------------------------------------------------------------------
>  23     Parameter #1 $objectOrClass of class WebFu\Reflection\ReflectionClass constructor expects class-string|object,
>          object|string given.

I'm not sure also what type of test would be good to prove API stability
I'l keep trying to proceed to the encapsulation, any suggestion is welcome

**Reply by veewee** [reply]
Instead of creating a wrapper for every object in ext-reflection, an alternative approach could be to provide functions to do 1 specific thing. These functions could return psl specific DTO's that can be used to request additional data:

* `Reflect\class_methods($classOrObject) : list<MethodDTO>`
* `Reflect\class_properties($classOrObject) : list<PropertyDTO>`
* `Reflect\method_arguments(MethodDTO $method) : list<ArgumentDTO>;`

By using DTO's instead of the actual reflection classes, we can make sure it is all type-safe and that the implementation stays within PSL and therefore does not leak back to PHP's reflection classes.

That way we could gradually build up the reflection library based on what people use / need instead of going full blown reflection replacement right away. It allows us to learn a bit more about the pitfalls.

An additional benefit of this approach could be the composability between other parts of PSL and we don't need to implement things like [flagged filters](https://www.php.net/manual/en/reflectionclass.getmethods.php).

For example:

```php
$privateMethods = Vec\filter(
    Reflect\class_methods($class),
    static fn (MethodDto $method): bool => $method->visibility === Visibility::PRIVATE
);
```

(of course we could add shortcut functions like `Reflect\private_class_methods` if it makes sense)


Another example of composability:

```php
$allDistinctClassMethodsAttributeNames = Vec\unique(
    Vec\map(
        Vec\flatten(
            Reflect\class_methods($class),
            static fn (MethodDto $method): array => Reflect\method_attrbutes($method)
        ),
        static fn (AttributeDto $attribute): string => $attribute->name
    )
);
```

I'm wondering though, how many work the \ReflectionClass e.g. does during construction.
At first sight, that seems to be very small amount of tasks:

* https://github.com/php/php-src/blob/master/ext/reflection/php_reflection.c#L3953

Since PHP's reflection is quite fast - it could be an idea not to keep track of \ReflectionClass instances and instead create new once when needed. A benchmark must be done before concluding this approach of course!
(If it's to slow, we could keep an internal lookup table, maybe weakref based - dunno if it's a good idea :) )

**Reply by web-fu** [reply]
Atomic functions are a great idea, I wouldn't exclude we can have wrapping classes and atomic functions working together. 
I noticed that Reflection leads to recursive definitions, ie: 
```$reflectionClass->getReflectionConstant()->getDeclatingClass()```
should return `$reflectionClass`
Maybe using a singleton to store explored elements could help performance

**Reply by azjezz** [reply]
> Maybe using a singleton to store explored elements could help performance

I don't think that would be necessary, native reflections, unlike many people assume, are super fast, so eating up memory to avoid reconstruction is not worth it.

If a project heavily depends on reflections, they should instead use native reflections or BetterReflections, the point of most Psl components is to cover the most common use cases, not act as a total replacement

**Reply by web-fu** [reply]
> I don't think that would be necessary, native reflections, unlike many people assume, are super fast, so eating up memory to avoid reconstruction is not worth it.

Can you suggest some benchmark tool? I tried singleton strategy on my PoC repo, I'm curious about results.

I think the most memory / runtime expensive thing is to get types from annotations.
You need to check for use statements (that requires reading the file) and resolve aliased types.

**Reply by veewee** [reply]
@web-fu you could use https://github.com/phpbench/phpbench

**Reply by web-fu** [reply]
@veewee I tried to run phpbench and the difference with or without singleton is 0.1 microsecond (10.000 revolution) for the function that read class file, tokenize it and return declared `use` statements.
This is, by far, the more complex operation in the entire wrapping library. 
Last commit here: https://github.com/web-fu/reflection/tree/e59f0cdc72ea949f961719dd37ab588916e43f8c
With this small difference I think i gonna drop singletons (less code = better code)

**Reply by veewee** [reply]
Nice! That would make it possible to create a set of atomic functions then. 
Nevertheless, the approach you took in your repo looks fine as well.

What do you need to parse the use statements for? Looks a bit add to manually parse the file in order to get use statement information.

**Reply by web-fu** [reply]
> What do you need to parse the use statements for? Looks a bit add to manually parse the file in order to get use statement information.

In reality is not so difficult. I found a gist: https://gist.github.com/Zeronights/7b7d90fcf8d4daf9db0c and I adapted it a little.
The core function is `token_get_all()`, a built-in tokenizer: https://www.php.net/manual/en/function.token-get-all.php
```
$reflectionClass = self::createReflectionClass($className);

if (!$reflectionClass->getFileName()) {
    throw new ReflectionException('Unable to retrieve filename for class ' . $className);
}

$source = file_get_contents($reflectionClass->getFileName());
if (!$source) {
    throw new ReflectionException('Could not open file ' . $reflectionClass->getFileName());
}

$tokens = token_get_all($source);

$builtNamespace = '';
$useStatements = [];
$class = '';
$as = '';
$buildingNamespace = false;
$buildingUse = false;
$buildingAs = false;

foreach ($tokens as $token) {
    if (is_array($token)) {
        if ($token[0] === T_NAMESPACE) {
            $buildingNamespace = true;
        }

        if ($token[0] === T_USE) {
            $buildingUse = true;
        }

        if ($token[0] === T_AS) {
            $buildingAs = true;
        }
    }

    if ($token === ';') {
        $buildingNamespace = false;
        $buildingUse = false;
        $buildingAs = false;

        if ($builtNamespace) {
            $builtNamespace = trim($builtNamespace);
        }

        if ($class) {
            /** @var class-string $class */
            $class = trim($class);
            $as = $as ? trim($as) : $class;
            $useStatements[] = new ReflectionUseStatement($class, $as);
            $class = '';
            $as = '';
        }
    }

    if ($buildingNamespace) {
        if ($token[0] === T_NAMESPACE) {
            continue;
        }

        $builtNamespace .= $token[1];

        continue;
    }

    if ($buildingUse) {
        if ($token[0] === T_USE) {
            continue;
        }

        if ($token[0] === T_AS) {
            continue;
        }

        if (! $buildingAs) {
            $class .= $token[1];
        } else {
            $as .= $token[1];
        }
    }
}
```
I am not sure 100% that is correct, I need more test cases and code coverage is low. 
I would like to add mutation testing, but first I need to learn how to use that tool

**Reply by veewee** [reply]
I was wondering more about the use-case of why you need to parse it manually.
What do you need that information for?

The reason I ask: a file could contain multiple classes - which are sharing or not sharing the same imports.
So this information is rather a reflection of a file than reflection info of a class.
Doesn't the reflected information give you the fully qualified class name of the imported symbol?

**Reply by azjezz** [reply]
classes could also exist outside of files, e.g generated using `eval()`.

i don't see why we can't just use builtin reflections.

**Reply by web-fu** [reply]
> I was wondering more about the use-case of why you need to parse it manually. What do you need that information for?
> 
> The reason I ask: a file could contain multiple classes - which are sharing or not sharing the same imports. So this information is rather a reflection of a file than reflection info of a class. Doesn't the reflected information give you the fully qualified class name of the imported symbol?

@veewee 
You need use statements because sometime you need to resolve a partial class name.
Imagine you want to support DocBlock Types:

```
use My\Own\Namespace\Foo as Foo;

/**
 * @template Test
 */
class ClassWithDocComments
{
    /** @var Foo[] */
    private array $useStatementDocComment;
}
```

What is the docType of `$useStatementDocComment`?

If you use standard `ReflectionProperty::getDocComment()` you can guess the type is "array of `Foo`", but in reality you need full FQSEN: My\Own\Namespace\Foo

@azjezz 
> classes could also exist outside of files, e.g generated using `eval()`.
> 
> i don't see why we can't just use builtin reflections.

Most of the job can be done by builtin reflection (apart the annoying use of `false` instead of `null` and `null` instead of `undefined`) but there are some corner cases (such as typeresolving) that would benefit of a minimal improvement.

https://github.com/Roave/BetterReflection and https://docs.phpdoc.org/ solve these problems but they are much bigger libriaries that are discouraged to be used in runtime applications.

**Reply by veewee** [reply]
Just wanted to mention that I was in a need for runtime reflections as well.
I took a slightly different approach than @web-fu that is more in line with the comment in https://github.com/azjezz/psl/discussions/402#discussioncomment-5166510.

It's still very basic at this moment:
https://github.com/veewee/reflecta/blob/main/docs/reflect.md

Because it's a part of a bigger package to enable making your own `optics`. The reflection part is all I needed at the moment.
https://github.com/veewee/reflecta

If this is an approach people like, I'm in for adding this to PSL instead.
However, I don't think that parsing PHP tokens and docblocks would be some steps too far for this package.
Instead you could rather use something like https://github.com/phpstan/phpdoc-parser instead maybe.

**Reply by web-fu** [reply]
Your implementation look similar to my reflection - related project: https://github.com/web-fu/anymapper, https://github.com/web-fu/php-dot-notation.
I'll try phpstandoc-parser and check if I can have better results.

**Reply by veewee** [reply]
Those look nice as well, cool :)

I'dd say mine rather aims to act as the base components that can be used by implementations like mappers, dot notation wrappers, ... (Like yours). Other implementations could be encoders, validators, ...

**Reply by web-fu** [reply]
Last days I updated my reflection library adding support from PHP 8.0 to 8.3, enums and others things.
Documentation is still lacking, but test coverage is quite good.
The only way I see now is to improve further is to test against real-life scenarios

**Reply by veewee** [reply]
Nice to hear. Quickly skimmed around it:
In reflecta, I was also [taking the path](https://github.com/veewee/reflecta/pull/3/files) of decorating the existing reflection classes. But i'm reluctant to take that path since it feels a bit verbose to me. I'm gonna close that PR and start a new one later:

When using reflection for runtime operations, you mostly just want to perform some very specific actions : take the value of a prop, list all contants, see if a method exists, get all class/property/method attribute ... So to me it makes sense to take a more functional approach which focusses on doing one very specific action per function call. Instead of decoration what already exists, I want to focus on specific actions one would want to apply.

For example:

```php
$values = [];
$publicProps = object_properties($object, property_visibility(Visibility::PUBLIC));

foreach ($publicProps as $propName) {
     $values[$propName] = property_get($object, $propName);
}
```

Non of these function leak any of the reflection internals. (altough you could create a custom predicate function which would leak the `\ReflectionProperty` instance in order to make decisions).


This API has some interesting advantages to me:

* It is fully task based : You don't need people to know about everything you can do with reflection, rather how to use it.
* It keeps the hard reflection part away from the end-users.
* You can pass in any custom predicate to filter the result - instead of working with the reflection flags. (although this kinda leaks the reflection part)
* You can't pass both string|object as param. Downside : We need both `object_properties` and `class_properties` - which are in big lines the same - but with small nuances like dynamic properties.
* Take for [example](https://github.com/veewee/reflecta/blob/main/src/Reflect/object_attributes.php) : `object_attributes($object, $predicate)` : you can return the initialized object attributes instead of a `ReflectionAttribute` making it all easy to use for the end users without having to deal with reflection theirselves.
* Error handling: a lot of reflection functions result in a ReflectionExceptions which all need to be dealt with. This becomes a bit verbose, but it's all contained in small functions. (Example [1](https://github.com/veewee/reflecta/blob/main/src/Reflect/property_set.php), [2](https://github.com/veewee/reflecta/blob/main/src/Reflect/Internal/reflect_property.php))
* It's easier to create an immutable API this way ([example](https://github.com/veewee/reflecta/blob/main/src/Reflect/property_set.php))


I'm just typing out my thoughts here. Feel fee to counter them : In the end we're all just looking for a good way to deal with reflection :)

**Reply by web-fu** [reply]
Since most of the Reflection application are atomic you may be right.
If you browse my encapsulation most of the time I return the original result (or the wrapped version of the original result)
My phylosophy writing my library was "keep as much similar to standard PHP reflection as possible" and "keep as strict type as possibile"
Reflection can be hard for newbie, and I wanted to keep the learning curve as flat as possible.
I don't think there is an absolute best in our approach.
I wrote a toolbox that you can use for example to build a task-based, functional, immutable API
I'll try to rewrite some of your examples and see what happens
